pkg/fqdn/dnspoller.go

```diff
@@ -0,0 +1,381 @@
+}
+
+// StopPollForDNSName runs the bookkeeping to remove each api.Rule from the
+// dnsPollTargets. When a target is empty, it is removed from the DNS names
+// that are polled.
+// Note: rule deletion in policy.Repository is by label, where at least that
+// many labels must match. This means our IP modified rule will be deleted
+// correctly, and no action is needed here.
+func (poller *DNSPoller) StopPollForDNSName(sourceRules []*api.Rule) {
+	for _, sourceRule := range sourceRules {
+		// skip untagged rules, nothing to do
+		if !sourceRule.Labels.Has(uuidLabelSearchKey) {
+			continue
+		}
+
+		uuid := getUUIDFromRuleLabels(sourceRule)
+		noLongerPolledDNSNames := poller.removeRule(uuid, sourceRule)
+		log.WithFields(logrus.Fields{
+			"noLongerPolled": noLongerPolledDNSNames,
+		}).Debug("Removed FQDN from poll list")
+	}
+}
+
+// LookupUpdateDNS runs a DNS lookup for each stored DNS name, storing updates,
+// and then emits regenerated policy rules.
+// The general steps are:
+// 1- take a snapshot of DNS names to lookup from poller, into dnsNamesToPoll
+// 2- Do a DNS lookup for each DNS name (map key) in poller via LookupDNSNames
+// 3- Update IPs for each dnsName in poller. If the IPs have changed for the
+// name, store which rules must be updated in rulesToUpdate. This is a set and
+// is deduped
+// 4- For each rule in rulesToUpdate, generate a new policy rule with IPs
+// 5- If we have any rules to update, emit them with AddGeneratedRules
+func (poller *DNSPoller) LookupUpdateDNS() error {
+	// Collect the DNS names that need lookups. This avoids locking
+	// poller during lookups.
+	dnsNamesToPoll := poller.GetDNSNames()
+
+	// lookup the DNS names. Names with failures will not be updated (and we
+	// will use the most recent data below)
+	updatedDNSIPs, errorDNSNames := poller.config.LookupDNSNames(dnsNamesToPoll)
+	for dnsName, err := range errorDNSNames {
+		log.WithError(err).WithField("fqdn", dnsName).Warn("Cannot resolve FQDN")
+	}
+
+	// Update IPs in poller
+	rulesToUpdate, updatedDNSNames := poller.UpdateDNSIPs(updatedDNSIPs)
+	for dnsName, IPs := range updatedDNSNames {
+		log.WithFields(logrus.Fields{
+			"fqdn": dnsName,
+			"IPs":  IPs,
+		}).Debug("Updated FQDN with new IPs")
+	}
+
+	// Generate a new rule for each sourceRule that needs an update.
+	var generatedRules []*api.Rule
+	for sourceRule := range rulesToUpdate {
+		newRule, err := generateRuleFromSource(sourceRule, updatedDNSNames)
+		if err != nil {
+			// TODO: make this more specific
+			log.WithError(err).Warn("Error generating policy rule with IPs for ToFQDN rule")
+		}
+
+		generatedRules = append(generatedRules, newRule)
+	}
+
+	// no rules to add, do not call PolicyAdd below
+	if len(generatedRules) == 0 {
+		return nil
+	}
+
+	// emit the new rules
+	return poller.config.AddGeneratedRules(generatedRules)
+}
+
+// GetDNSNames returns a snapshot of the DNS names in DNSPoller
+func (m *DNSPoller) GetDNSNames() (dnsNames []string) {
+	m.Lock()
```

Comment:
How frequently is the GetDNSNames() called? Is it worth changing to RLock() and RUnlock()?